Potentially redundant check for zero alignment in generated assembly

Consider the following code with custom DSTs:

```rust
pub trait My {
    fn f(&self) -> i32;
}

pub struct Wrapper<T: ?Sized>(i8, T);

pub struct My1<T: My + ?Sized> {
    field: i8,
    my: Wrapper<T>,
}

type DynMy1 = My1<dyn My>;

pub fn run(d: &DynMy1) -> &Wrapper<dyn My> {
    &d.my
}
```

The generated assembly for x86_64 is as follows (with my comments added):

```asm
example::run:
  mov     rdx, rsi                   ; Copy vtable pointer into that of result
  mov     rcx, qword ptr [rsi + 16]  ; Load alignment into rcx
  test    rcx, rcx                   ; Check if the alignment is non-zero (?)
  mov     eax, 1
  cmovne  rax, rcx                   ; If the alignment is zero, then the offset is 1, otherwise it's rcx
  add     rax, rdi                   ; Add offset to base pointer and store into the base pointer of result
  ret
```

This can be seen on [Godbolt](https://godbolt.org/z/hYM8cz8Kr).

Why does the compiler check for zero alignment? It seems that it's impossible, as the alignment cannot be zero.

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0 (09c42c458 2021-10-18)
binary: rustc
commit-hash: 09c42c45858d5f3aedfa670698275303a3d19afa
commit-date: 2021-10-18
host: x86_64-unknown-linux-gnu
release: 1.56.0
LLVM version: 13.0.0
```

Adding
```rust
fn main() {
    println!("{}", std::mem::align_of::<Wrapper<dyn My>>());
}
```
gets me
```rust
error[E0277]: the size for values of type `dyn My` cannot be known at compilation time
   --> <source>:19:20
    |
19  |     println!("{}", std::mem::align_of::<Wrapper<dyn My>>());
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: within `Wrapper<dyn My>`, the trait `Sized` is not implemented for `dyn My`
note: required because it appears within the type `Wrapper<dyn My>`
   --> <source>:5:12
    |
5   | pub struct Wrapper<T: ?Sized>(i8, T);
    |            ^^^^^^^
note: required by a bound in `align_of`
```

So, at least as far as the compiler is concerned, there is no obvious reason it could not be a zero-sized type.
I could believe this is a missed optimization, where the compiler could infer that because of the i8 field, it has to be aligned to at least 1, but I do not believe it is a bug. It seems to satisfy the current reasoning around alignment and DSTs.

--- a/compiler/rustc_codegen_ssa/src/glue.rs
+++ b/compiler/rustc_codegen_ssa/src/glue.rs
@@ -6,6 +6,7 @@ use crate::common::IntPredicate;
 use crate::meth;
 use crate::traits::*;
 use rustc_middle::ty::{self, Ty};
+use rustc_target::abi::WrappingRange;
 
 pub fn size_and_align_of_dst<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     bx: &mut Bx,
@@ -21,14 +22,17 @@ pub fn size_and_align_of_dst<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     }
     match t.kind() {
         ty::Dynamic(..) => {
-            // load size/align from vtable
+            // Load size/align from vtable.
             let vtable = info.unwrap();
-            (
-                meth::VirtualIndex::from_index(ty::COMMON_VTABLE_ENTRIES_SIZE)
-                    .get_usize(bx, vtable),
-                meth::VirtualIndex::from_index(ty::COMMON_VTABLE_ENTRIES_ALIGN)
-                    .get_usize(bx, vtable),
-            )
+            let size = meth::VirtualIndex::from_index(ty::COMMON_VTABLE_ENTRIES_SIZE)
+                .get_usize(bx, vtable);
+            let align = meth::VirtualIndex::from_index(ty::COMMON_VTABLE_ENTRIES_ALIGN)
+                .get_usize(bx, vtable);
+
+            // Alignment is always nonzero.
+            bx.range_metadata(align, WrappingRange { start: 1, end: !0 });
+
+            (size, align)
         }
         ty::Slice(_) | ty::Str => {
             let unit = layout.field(bx, 0);

--- a/src/test/codegen/dst-vtable-align-nonzero.rs
+++ b/src/test/codegen/dst-vtable-align-nonzero.rs
@@ -1,3 +1,5 @@
+// compile-flags: -O
+
 #![crate_type = "lib"]
 
 // This test checks that we annotate alignment loads from vtables with nonzero range metadata,

--- a/src/test/codegen/dst-vtable-align-nonzero.rs
+++ b/src/test/codegen/dst-vtable-align-nonzero.rs
@@ -1,0 +1,43 @@
+#![crate_type = "lib"]
+
+// This test checks that we annotate alignment loads from vtables with nonzero range metadata,
+// and that this allows LLVM to eliminate redundant `align >= 1` checks.
+
+pub trait Trait {
+    fn f(&self);
+}
+
+pub struct WrapperWithAlign1<T: ?Sized> { x: u8, y: T }
+
+pub struct WrapperWithAlign2<T: ?Sized> { x: u16, y: T }
+
+pub struct Struct<W: ?Sized> {
+    _field: i8,
+    dst: W,
+}
+
+// CHECK-LABEL: @eliminates_runtime_check_when_align_1
+#[no_mangle]
+pub fn eliminates_runtime_check_when_align_1(
+    x: &Struct<WrapperWithAlign1<dyn Trait>>
+) -> &WrapperWithAlign1<dyn Trait> {
+    // CHECK: load [[USIZE:i[0-9]+]], {{.+}} !range [[RANGE_META:![0-9]+]]
+    // CHECK-NOT: icmp
+    // CHECK-NOT: select
+    // CHECK: ret
+    &x.dst
+}
+
+// CHECK-LABEL: @does_not_eliminate_runtime_check_when_align_2
+#[no_mangle]
+pub fn does_not_eliminate_runtime_check_when_align_2(
+    x: &Struct<WrapperWithAlign2<dyn Trait>>
+) -> &WrapperWithAlign2<dyn Trait> {
+    // CHECK: [[X0:%[0-9]+]] = load [[USIZE]], {{.+}} !range [[RANGE_META]]
+    // CHECK: [[X1:%[0-9]+]] = icmp {{.+}} [[X0]]
+    // CHECK: [[X2:%[0-9]+]] = select {{.+}} [[X1]]
+    // CHECK: ret
+    &x.dst
+}
+
+// CHECK: [[RANGE_META]] = !{[[USIZE]] 1, [[USIZE]] 0}

